src/login.ts

```diff
@@ -227,6 +228,12 @@ const states = [
       debug("Typing password");
       await page.keyboard.type(password);
 
+      // federation with on-prem ADFS causes an HTTP basic auth prompt inside a redirect here
```

Comment:
Is there a good way to programmatically know if there is an HTTP basic window open with Puppeteer? That way we can fire this off only if necessary. If not, will this be bad for security?

Reply:
I spent several hours last night @kennethaasan looking for this (totally new to pupetteer).

I identified two possible methods:
1) Create a state that recognises an auth prompt in `const states` using a selector
The only issue I found was there are no elements on an auth window so I can't find a way to do that. 

2) During the 'State not found' loop (which will attempt several times so as to not exit the application if the page is simply slow to load etc.), add an auth window check/push credentials (assuming no security issue... why would there be?)

```
# ~L765
          if (foundState) {
            totalUnrecognizedDelay = 0;
          } else {
            <checkForAuthWindow>
            debug("State not recognized!");
```


3) On `frameScheduledNavigation`, test the `page.url` contains `adfs.*` 
I tried this however, even though `page.url` returns a string, it did not have the `contains` method so I was a bit stumped.

Reply:
```
// ADFS
let username: string;
let password: string;
let adfs: boolean;
```

```
          if (foundState) {
            totalUnrecognizedDelay = 0;
          } else {
            debug("State not recognized!");

            if (adfs) {
              debug("Authenticating with ADFS")
              await page.authenticate({
                username,
                password,
              });
            }

            if (totalUnrecognizedDelay > MAX_UNRECOGNIZED_PAGE_DELAY) {
```